// angular-intl/esm5/src/translate.pipe.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Pipe } from '@angular/core';
import { TranslateService } from './translate.service';
import { Subject } from 'rxjs';
import { filter, takeUntil } from 'rxjs/operators';
var TranslatePipe = /** @class */ (function () {
    function TranslatePipe(translateService) {
        this.translateService = translateService;
        this.translation = '';
        this.unsubscribe = new Subject();
        this.translationLoaded$ = translateService.translationsLoaded.pipe(filter(Boolean), takeUntil(this.unsubscribe));
    }
    /**
     * @param {?} val
     * @param {?} args
     * @return {?}
     */
    TranslatePipe.prototype.transform = /**
     * @param {?} val
     * @param {?} args
     * @return {?}
     */
    function (val, args) {
        var _this = this;
        this.translationLoaded$.subscribe(function () {
            var /** @type {?} */ readValue = val ? _this.translateService.read(val, args) : '';
            _this.translation = readValue === val ? _this.translation : readValue;
        });
        return this.translation;
    };
    /**
     * @return {?}
     */
    TranslatePipe.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.unsubscribe.next();
        this.unsubscribe.complete();
    };
    TranslatePipe.decorators = [
        { type: Pipe, args: [{
                    name: 'translate',
                    pure: false
                },] }
    ];
    /** @nocollapse */
    TranslatePipe.ctorParameters = function () { return [
        { type: TranslateService, },
    ]; };
    return TranslatePipe;
}());
export { TranslatePipe };
function TranslatePipe_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    TranslatePipe.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    TranslatePipe.ctorParameters;
    /** @type {?} */
    TranslatePipe.prototype.translation;
    /** @type {?} */
    TranslatePipe.prototype.translationLoaded$;
    /** @type {?} */
    TranslatePipe.prototype.unsubscribe;
    /** @type {?} */
    TranslatePipe.prototype.translateService;
}
