// angular-intl/esm2015/src/translate.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { NgModule } from '@angular/core';
import { HttpClientModule } from '@angular/common/http';
import { TranslateDirective } from './translate.directive';
import { TranslatePipe } from './translate.pipe';
import { TranslateService } from './translate.service';
import { LoaderService } from './loader.service';
import { CONFIG } from './types/config.token';
export class TranslateModule {
    /**
     * @param {?=} config
     * @return {?}
     */
    static forRoot(config = {}) {
        return {
            ngModule: TranslateModule,
            providers: [
                TranslateService,
                LoaderService,
                { provide: CONFIG, useValue: config }
            ]
        };
    }
    ;
    /**
     * @return {?}
     */
    static forChild() {
        return {
            ngModule: TranslateModule
        };
    }
    ;
}
TranslateModule.decorators = [
    { type: NgModule, args: [{
                imports: [HttpClientModule],
                declarations: [
                    TranslateDirective,
                    TranslatePipe
                ],
                exports: [
                    TranslateDirective,
                    TranslatePipe
                ]
            },] }
];
function TranslateModule_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    TranslateModule.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    TranslateModule.ctorParameters;
}
